op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+  switch x := event.Data.(type) {
+  case *NewPayload:
+      // 1) update some rollup state: ...
+      // 2) and propagate to the engine deriver
+      r.engine.OnEvent(ctx, event)
+  case *L1Change:
+      r.pipeline.OnEvent(ctx, event)
+  default:
+      r.pipeline.OnEvent(ctx, event) // can forward anything to a fallback
+  }
+}
+```
+
+Or, a deriver can be given access to the root-level Deriver, to surface new event data to:
+```go
+type Pipeline struct {
+    root Deriver
+}
+
+func (p *Pipeline) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+    case *L1Change:
+        // generate block attributes and give them to the unsafe-block processor to extend the chain
+        // (or forward to next deriver if already known attributes)
+        p.root.OnEvent(ctx, &Event{Resource: UnsafeL2, Data: AttributesEvent{attributes}})
+    }
+}
+```
+
+#### Deriver design principles
+
+Custom stacks like various Alt-DA implementations can wrap around any existing derivers,
+extend with new events, possibly re-using standard OP-Plasma events, for additional de-duplication of alt-DA code.
+
+To encourage composability, and reduce code, events should be documented and small in scope.
+I.e. create an event for a deriver to turn attributes in an unsafe block,
+and follow that up with events to try to promote the block safety, suggest it for finality, and persisting it to the safe head DB.
+rather than tightly coupling the four acts: (1) block processing, (2) updating of the safe-head, (3) remembering the finality data all at once, (4) persisting the safe head change.
+
+
+#### Threads and parallel derivers
+
+Stacks can define a single `OnEvent` handler, and by using `Resource`, still execute on parallel threads, if desired.
+Note that things can be entirely synchronous, if the caller ignores the resource preference, and just runs one thing at a time. This can support fault-proofs, reducing the custom code (no fault-proof specific driver logic, just a synchronous loop of calling the deriver of choice).
+
+To get the parallelism, the `Driver`, which itself can be the absolute root `Deriver`,
+can create a new go routine for each new `Resource` type, when first seen.
+```go
+worker, ok := driver.workers[event.Resource]
+if !ok {
+    worker = ... // spawn new worker
+    driver.workers[event.Resource] = worker // keep it around
+}
+worker.OnEvent(ctx, event)
+```
+
+#### Instrumentation
+
+A `InstrumentedDeriver` can be introduced, as a shim for `OnEvent`, wrapping an underlying `Deriver`,
+to provide a default set of instrumentation for all current future event handling:
+- Measure the execution-time per event type.
+- Track utilization per resource.
+- Trace/debug log the results and timeouts.
+
+This can help identify bottlenecks, does not have to be rewritten with each op-stack mod/extension,
+and fits it all in just a few reusable/aggregate metrics charts.
+
+### Spec changes
+
+In the specs we enshrine:
```

Comment:
This document is important for describing how the op-node works & how to modify it, but I don't think we should enshrine implementation in the spec.